File: java/src/VolatilePublication.java
/**
 * Demonstrates safe publication using `volatile`.
 *
 * The writer publishes `payload` (a plain field) and then flips a
 * `volatile` flag. Because a volatile write is a *release* and a
 * volatile read of the same field is the matching *acquire*, the JMM
 * guarantees that once the reader observes `ready == true`, it also
 * observes the plain write to `payload` that happened before it.
 *
 * Try removing `volatile` and this program can (on some JITs/CPUs,
 * over enough iterations) print 0 instead of 42 -- or spin forever,
 * since the reader may never observe the flag flip at all.
 */
public class VolatilePublication {

    private static volatile boolean ready = false;
    private static int payload;

    private static void produce() {
        payload = 42;   // (1) plain write
        ready = true;   // (2) volatile write -- release
    }

    private static void consume() throws InterruptedException {
        while (!ready) {   // (3) volatile read -- acquire
            Thread.onSpinWait();
        }
        System.out.println("Consumer saw payload = " + payload);
    }

    public static void main(String[] args) throws InterruptedException {
        Thread consumer = new Thread(() -> {
            try {
                consume();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });
        consumer.start();

        // Give the consumer a head start spinning on `ready`.
        Thread.sleep(50);
        produce();

        consumer.join();
    }
}

File: cpp/src/volatile_publication.cpp
// Demonstrates safe publication using std::atomic release/acquire.
//
// NOTE: C++ `volatile` is NOT a threading primitive -- it only stops the
// compiler from optimizing away accesses to memory-mapped I/O. Using it
// here for cross-thread visibility (as one might, coming from Java) is a
// classic bug. The correct tool is std::atomic with explicit ordering.
//
// The writer publishes `payload` (a plain int) and then stores `true`
// into an atomic<bool> using memory_order_release. The reader spins on
// an memory_order_acquire load of the same flag. That release/acquire
// pair guarantees the plain write to `payload` is visible to the reader
// once it observes `ready == true` -- exactly like Java's volatile.

#include <atomic>
#include <thread>
#include <chrono>
#include <iostream>

std::atomic<bool> ready{false};
int payload = 0;

void produce() {
    payload = 42;                                   // (1) plain write
    ready.store(true, std::memory_order_release);   // (2) release
}

void consume() {
    while (!ready.load(std::memory_order_acquire)) { // (3) acquire
        std::this_thread::yield();
    }
    std::cout << "Consumer saw payload = " << payload << "\n";
}

int main() {
    std::thread consumer(consume);
    std::this_thread::sleep_for(std::chrono::milliseconds(50)); // head start
    produce();
    consumer.join();
    return 0;
}
